api/repositories/shared.go

```diff
@@ -1,15 +1,56 @@
 package repositories
 
 import (
+	"context"
 	"errors"
+	"fmt"
 
 	"code.cloudfoundry.org/cf-k8s-controllers/api/authorization"
 	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
 
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+var CFAppGVR = schema.GroupVersionResource{
+	Group:    "workloads.cloudfoundry.org",
+	Version:  "v1alpha1",
+	Resource: "cfapps",
+}
+
+type NamespaceRetriever struct {
+	client dynamic.Interface
+}
+
+func NewNamespaceRetriver(client dynamic.Interface) NamespaceRetriever {
+	return NamespaceRetriever{
+		client: client,
+	}
+}
+
+func (n NamespaceRetriever) GetNamespacedResourceNamespace(ctx context.Context, resourceGUID string, gvr schema.GroupVersionResource) (string, error) {
+	opts := metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", resourceGUID),
+	}
+	list, err := n.client.Resource(gvr).List(ctx, opts)
+	if err != nil {
+		return "", fmt.Errorf("failed to list %v", gvr)
+	}
+
+	if len(list.Items) == 0 {
+		return "", NewNotFoundError(gvr.Resource, nil)
+	}
+	if len(list.Items) > 1 {
+		return "", fmt.Errorf("get-%s duplicate records exist", gvr.Resource)
+	}
+
+	metadata := list.Items[0].Object["metadata"].(map[string]interface{})
+
+	return metadata["namespace"].(string), nil
```

Comment:
Maybe we should check if namespace == "", which would mean the object is not namespace scoped, in which case we can return an error? I guess this is only possible because of programmer error, but still it will be good to catch it early.

Reply:
I like this sort of error checking. Thanks for the input.